### DISCUSSION TITLE
Firmware Architecture

### DISCUSSION THREAD

**Discussion by longevity-software** [VolksEEG/VolksEEG#11] [Firmware]
### Firmware Architecture Overview
I have included a plantuml diagram of what I am initially proposing as the firmware architecture, this includes the following main features:

- A Core (platform agnostic) set of functionality, which will be at the heart of the firmware.
- A Driver Layer which will abstract specific pieces of hardware e.g the ADS1299.
- A Hardware Abstraction Layer (HAL) which abstracts the generic hardware peripherals.
- An OS (and possibly an OS abstraction).

![image](https://user-images.githubusercontent.com/38405059/140817679-c8ff122d-014f-4165-b19c-3d7add41b7ee.png)

Lets start a discussion around this and get everyone's thoughts.

**Comment by alcohen** [comment]
First off, it looks sophisticated and well thought through, thanks Graham!

I'm not a firmware expert, so my input might not be the most useful of inputs.

I get a basic idea of what's going on, but I think I'd need to walk through how the data flow during various operations.

I do have another higher-level question that I'm trying to formulate - I'll share it once I do.

**Comment by stevebranam** [comment]
That looks good to me. It's similar to what I was thinking, with the advantage that it's bolstered by @longevity-software's actual experience with such products! My thoughts were a pretty generic starting point.

**Comment by harry-brown** [comment]
Looks great @longevity-software! I'm guessing all of the data flowing directly between ADS1299Driver and ComsParser is the control/status of the ADS?

**Reply by longevity-software** [reply]
Correct, the link between ComsParser and ADS1299Driver would be just configuration and status. 
The link between ComsParser and DataCenter would be EEG and impedance data.

**Reply by longevity-software** [reply]
I had overlooked impedance measurements in the diagram above, it could either be part of DataCenter as mentioned above, or could be it's own module, it's own module probably makes more sense unless edf files contain impedance information.
